Se agrega logica inicial para el submit final de la encuesta

patriciotella opened the pull request.

patriciotella committed:
Se agrega logica inicial para el submit final de la encuesta

jcalvento commented on app/controllers/surveys_controller.rb:
```diff
@@ -1,7 +1,17 @@
-    rand
-    render json: { message: 'Survey successfully submitted' }, status: :ok
+    success_message = ''
+    params[:subjects].each do |subject|
```
Acá no tenes que usar las comisiones en vez de la materia?

patriciotella replied:
yo recibo desde el front la colleccion de materias y la comision (o la opcion) que eligio el alumno

jcalvento commented on app/controllers/surveys_controller.rb:
```diff
@@ -1,7 +1,17 @@
+      current_subject = Subject.find_by(name: subject[:name])
+      student_answer = subject[:condition]
+      if student_answer == 'approve'
+      #   tenemos que asignarla a las materias aprobadas no como una respuesta
```
Esto ya lo hizo Roma en `SignupController#create`. cc @romanrizzi

romanrizzi replied:
Creo que el quería darle la oportunidad de que si el alumno se equivoco en el paso anterior, pueda corregirlo eligiendo la opción "Ya aprobe"

jcalvento commented on app/controllers/surveys_controller.rb:
```diff
@@ -22,4 +32,33 @@ def student_id
+        end
+        answer.reply_option = reply_option
+      else
+        chair = current_subject.subject_in_quarter.chairs.detect { |chair| chair.time == subject[:condition] }
```
Qué es la condition?

patriciotella replied:
le cambio el nombre por selected_chair

jcalvento commented on app/controllers/surveys_controller.rb:
```diff
@@ -22,4 +32,33 @@ def student_id
   end
+
+  def submit_answer(current_subject, student_answer, subject, success_message)
+    Answer.new.tap do |answer|
```
Movamos todo esto a otro lado

patriciotella replied:
ok, al helper del surver controller?

jcalvento commented on app/controllers/surveys_controller.rb:
```diff
@@ -22,4 +32,33 @@ def student_id
+      answer.student_id = params[:userId]
+      if !student_answer || student_answer == 'cant' || student_answer == ''
+        reply_option = ReplyOption.new.tap do |replyOption|
+          replyOption.value = 'No voy a cursar'
```
Este era el que dijimos que no íbamos a usar, guardemos sólo las que va a cursar y después vemos cómo hacer con los que tienen problemas con el horario

jcalvento reviewed (commented):
Clavale un par de tests

patriciotella committed:
Mejoras:

- Se extraen los metodos del controller a un helper.
- Se corrigen nombres de variables
- Se genera un json con las materias y el futuro link para edicion

romanrizzi commented on app/controllers/surveys_controller.rb:
```diff
@@ -1,7 +1,19 @@
+      if student_answer == 'approve'
+      #   tenemos que asignarla a las materias aprobadas no como una respuesta
+      else
+        success_message = submit_answer(current_subject, student_answer, subject, success_message)
```
para que guardamos esto en una variable, no se pisa constantemente en la iteración?

patriciotella replied:
Fijate que en el helper se le estan appedeando las respuesta y se pasa como variable

romanrizzi commented on app/controllers/surveys_controller.rb:
```diff
@@ -1,7 +1,19 @@
-    rand
-    render json: { message: 'Survey successfully submitted' }, status: :ok
+    success_message = []
+    params[:subjects].each do |subject|
```
Esto se podría hacer con un `reduce`

romanrizzi commented on app/helpers/surveys_helper.rb:
```diff
@@ -0,0 +1,35 @@
+      answer.save!
+    end
+    success_message
+  end
```
Esto metodo va en un service, no en un helper. Tenemos que separar la parte del dominio de la que corresponde a la presentación

patriciotella replied:
ok hago un survey_service y lo muevo ahi

romanrizzi commented on app/helpers/surveys_helper.rb:
```diff
@@ -0,0 +1,35 @@
+
+  def build_survey(survey_subjects)
+    survey_subjects.map { |subject| { name: subject.name, chairs: subject.chairs.map(&:time) } }
+  end
```
Hay que tener cuidado con las queries N+1 cuando nos traemos los chairs

patriciotella replied:
Esto lo habias hecho vos creo y con juli se fijaron eso antes de mergearlo a master, yo solo lo movi del controller

romanrizzi commented on app/helpers/surveys_helper.rb:
```diff
@@ -0,0 +1,35 @@
+  def submit_answer(current_subject, student_answer, subject, success_message)
+    Answer.new.tap do |answer|
+      answer.student_id = params[:userId]
+      if !student_answer || student_answer == 'cant' || student_answer == ''
```
Quizas podriamos delegar el estado en el objeto `Answer en lugar de un string`

patriciotella replied:
No tenes una answer, las estas creando con lo que recibis

patriciotella committed:
Cambios del pr:

- Se extraen los metodos del helper a un service.
- Se eliminan links provisorios para editar la encuesta, dejando un texto descriptivo hasta tener los links
-Fix rubocop offenses
-Fix spec

jcalvento commented on app/controllers/surveys_controller.rb:
```diff
@@ -1,7 +1,19 @@
+      current_subject = Subject.find_by(name: subject[:name])
+      student_answer = subject[:selectedChair]
+      if student_answer == 'approve'
+      #   tenemos que asignarla a las materias aprobadas no como una respuesta
```
Qué onda con esto?

jcalvento commented on app/services/survey_service.rb:
```diff
@@ -0,0 +1,41 @@
+  def submit_answer(current_subject, student_answer, subject, success_message)
+    Answer.new.tap do |answer|
+      answer.student_id = params[:userId]
+      if !student_answer || student_answer == 'cant' || student_answer == ''
```
cant no es que no puede en ese horario? Es distinto a "no voy a cursar". Y pensé que al final no íbamos a guardar las "no voy a cursar".

patriciotella committed:
Se validan los casos para los que el alumno no va a cursar ninguna materia, mostrandole solo el link de edicion
